### ISSUE TITLE
Update, document, or remove internal Schema.org template

### ISSUE THREAD

**Issue by ghost** [gohugoio/hugo#7431]
<pre>
$ hugo version
Hugo Static Site Generator v0.73.0-428907CC/extended darwin/amd64 BuildDate: 2020-06-23T16:37:47Z
</pre>

The current schema template does not define itemscope and itemtype, which are required for it to be valid. If you submit a page with the current meta tags to [W3.org Validator](https://validator.w3.org/) you will see errors:

`The itemprop attribute was specified, but the element is not a property of any item.`

If you run the HTML through [Google Structured Data Testing Tool](https://search.google.com/structured-data/testing-tool) no data will be parsed because it requires the above mentioned properties.

The wordCount meta entry in the template is only for the Article schema type, but if you set the type to Article in Google's tool, it will tell you it also requires "author", "headline", "publisher" which are not included in the template. The additional required fields that are included in the template, "datePublished" and "image", aren't required in the template.

In its current form, the template does not provide valid Schema.org structures. For users that are unfamiliar with Schema data structures it is adding HTML that won't validate and additional data that will not actually be parsed.

This template is mentioned, but not documented so users are not aware of the above issues. I feel documenting the template does little because the user still has to add several pieces of data to Params or frontmatter for anything meaningful to be added. The type of page is not as generic as OpenGraph (4 choices vs dozens in Schema)(ie "Article" is not always what I would consider "IsPage==True" to represent.)

**Comment by carlmjohnson** [comment]
On my site, I have some funky LD-JSON templates. Would that be a good replacement for the current internal metadata templates? Or do users need both?

**Comment by inwardmovement** [comment]
I agree, see [Understand how structured data works](https://developers.google.com/search/docs/guides/intro-structured-data)

**Comment by XhmikosR** [comment]
Perhaps after #3088 is sorted a new schema internal template could be added. For example https://codingnconcepts.com/hugo/structure-data-json-ld-hugo/

**Comment by rootwork** [comment]
@XhmikosR that issue doesn't seem to have made any significant progress in years. But thanks for that link to the structured data guide, that's very helpful!

**Comment by XhmikosR** [comment]
Yeah, but the authors situation needs to be standardized first... Right now the internal templates are broken when it comes to authors.

**Comment by rootwork** [comment]
Fair enough

**Comment by carlmjohnson** [comment]
I think this is still relevant.

**Comment by carlmjohnson** [comment]
I am starting to think pretty much all the internal templates should be replaced with blessed external modules.

**Comment by jmooring** [comment]
Among the three approaches listed in the issue title—update, document, or remove—additional documentation doesn't resolve the fact that the embedded schema partial generates invalid HTML:

<details>
<summary>rendered page</summary>

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>p1</title>
  
  <meta itemprop="name" content="p1">
  <meta itemprop="description" content="This is the description of p1.">
  <meta itemprop="wordCount" content="6">
  <meta itemprop="keywords" content="k1,k2">
</head>
<body>
  <p>This is the content of p1.</p>

</body>
</html>
```
</details>

<details>
<summary>validation results</summary>
<img width="812" height="430" alt="Image" src="https://github.com/user-attachments/assets/9ce5ad6f-c35c-482a-9bf9-e5b04df3fac6" />
</details>

My immediate preference is to deprecate and ultimately remove the embedded schema partial, thereby making site and theme authors responsible for schema generation. Although we can provide generic data, optimal (or even marginally acceptable) data requires custom page or site parameters. For example:

<details>
<summary>LD-JSON for content page</summary>

```json
{
  "@context": "https://schema.org",
  "@type": "Article",
  "author": {
    "@type": "Organization",
    "name": "Veriphor, LLC",
    "url": "https://www.veriphor.com/"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "name": "Veriphor, LLC"
  },
  "copyrightNotice": "© 2021 Veriphor, LLC.",
  "copyrightYear": 2021,
  "dateModified": "2025-04-12T07:16:59-07:00",
  "datePublished": "2021-12-13T17:12:00-08:00",
  "description": "Embed a Spotify widget in your Hugo site.",
  "headline": "Spotify widgets",
  "image": "https://www.veriphor.com/articles/spotify-widgets/card.png_11687639581236821446?fp=1834018779",
  "keywords": "hugo,shortcodes",
  "url": "https://www.veriphor.com/articles/spotify-widgets/",
  "wordCount": 1e3
}
```
</details>

<details>
<summary>LD-JSON for home page</summary>

```json
{
  "@context": "https://schema.org",
  "@type": "Organization",
  "address": {
    "@type": "PostalAddress",
    "addressLocality": "Seattle, Washington"
  },
  "description": "Veriphor, based in Seattle, provides software development and consulting services.",
  "email": "info@veriphor.com",
  "image": "https://www.veriphor.com/card.png_11687639581236821446?fp=554225494",
  "legalName": "Veriphor, LLC",
  "logo": "https://www.veriphor.com/images/branding/logo.png",
  "name": "Veriphor",
  "slogan": "We put the pieces together so you don't have to.",
  "telephone": "+1 206-457-2961",
  "url": "https://www.veriphor.com/"
}
```
</details>

<details>
<summary>LD-JSON for section page</summary>

```json
{
  "@context": "https://schema.org",
  "@type": "ItemList",
  "description": "A summarized list of articles.",
  "image": "https://www.veriphor.com/articles/card.png_11687639581236821446?fp=3896956768",
  "name": "Articles",
  "url": "https://www.veriphor.com/articles/"
}
```
</details>

If removing the embedded schema partial is not acceptable, I suggest we generate [LD-JSON](https://json-ld.org/) data. Details...

1. Create a map then pass through [`encoding.Jsonify`](https://gohugo.io/functions/encoding/jsonify/).
2. The map structure must be specific to each page kind. While `section`, `taxonomy`, and `term` pages can have the same map structure (`"@type" "ItemList"`), `home` and `page` pages must have their own structure.
3. Do not use custom page or site parameters. We've been down this path before, and it's not pretty.

**Comment by nisbet-hubbard** [comment]
> While section, taxonomy, and term pages can have the same map structure ("@type" "ItemList"), home and page pages must have their own structure.

On the second proposal, my only request is to expose these and other properties to be overridden via front matter, and fall back on the page type defaults when the author doesn't set a more specific value.

That wouldn't be necessary, though, if authors are expected to directly edit this partial anyway.
